src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java

```diff
@@ -0,0 +1,298 @@
+    private int maxComponents = DEFAULT_MAX_COMPONENTS;
+
+    /**
+     * Access modifiers of record definitions where the number
+     * of record components should be checked.
+     */
+    private AccessModifierOption[] accessModifiers = {
+        AccessModifierOption.PUBLIC,
+        AccessModifierOption.PROTECTED,
+        AccessModifierOption.PACKAGE,
+        AccessModifierOption.PRIVATE,
+    };
+
+    /**
+     * Setter to specify the maximum number of components allowed in record definition headers.
+     *
+     * @param value the maximum allowed.
+     */
+    public void setMaxComponents(int value) {
+        maxComponents = value;
+    }
+
+    /**
+     * Setter to access modifiers of record definitions where the number of record
+     * components should be checked.
+     *
+     * @param accessModifiers access modifiers of record definitions which should be checked.
+     */
+    public void setAccessModifiers(AccessModifierOption... accessModifiers) {
+        this.accessModifiers =
+                Arrays.copyOf(accessModifiers, accessModifiers.length);
+    }
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getAcceptableTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return new int[] {
+            TokenTypes.RECORD_DEF,
+        };
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.RECORD_DEF,
+        };
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        // Instantiate new component counter
+        final ComponentCounter counter = new ComponentCounter();
+
+        // Count record components in this record definition
+        countComponents(counter, ast);
+
+        // Check counter and report violations
+        checkCounter(counter, ast);
+    }
+
+    /**
+     * Determine the visibility modifier, then iterate through the record component list
+     * and increment the counter.
+     *
+     * @param counter the ComponentCounter instance for this check
+     * @param ast the ast to check
+     */
+    private static void countComponents(ComponentCounter counter, DetailAST ast) {
+        final DetailAST recordComponents =
+                ast.findFirstToken(TokenTypes.RECORD_COMPONENTS);
+
+        counter.setAccessModifierOption(getAccessModifier(ast));
+
+        TokenUtil.forEachChild(recordComponents,
+                TokenTypes.RECORD_COMPONENT_DEF, node -> counter.increment());
+
+    }
+
+    /**
+     * Check the counter and report violations.
+     *
+     * @param counter the component counter to check
+     * @param ast to report violations against.
+     */
+    private void checkCounter(ComponentCounter counter, DetailAST ast) {
+        if (matchAccessModifiers(counter.getAccessModifierOption())
+                && counter.getTotal() > maxComponents) {
+            log(ast, MSG_KEY, counter.getTotal(), maxComponents);
```

Comment:
Please move this to visitToken method, we should see logic in that method

Reply:
Done.

Reply:
Improved further.